### PR TITLE
Restrict product activations for PAYG instances

### PR DESCRIPTION
**Pull request by ikapelyukhin**

### CODE REVIEW

**Comment by suse-tests-fail** [comment]
Prophet reports failure.
https://ci.suse.de/job/scc-RMT-integration-tests/102739/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment.
Failed fast in project rmt on step docker-compose-run-tests. (Merged 028c5ac84bdfadaebdafb65a8a6965b1da53dd6b into 932ef69181d7ff536627caf00cfc4dc85f8e07de)

**Comment by suse-tests-fail** [comment]
Prophet reports failure.
https://ci.suse.de/job/scc-RMT-integration-tests/102740/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment.
Failed fast in project rmt on step docker-compose-run-tests. (Merged 028c5ac84bdfadaebdafb65a8a6965b1da53dd6b into bbff48aa07415f432017ea1c0b697448543b1e2c)

File: engines/instance_verification/lib/instance_verification/engine.rb [hunk 1]
```diff
@@ -30,9 +30,19 @@ def find_product
...
+        def verify_product_activation
           product = find_product
-          return unless product.base?
+
+          unless product.base?
+            return if product.free?
+            base_product = @system.products.where(product_type: :base).first
+            if base_product.identifier != 'SLES_SAP'
+              raise InstanceVerification::Exception.new(
+                'The product is not available for this image'
```
Comment (rjschwei) [inline engines/instance_verification/lib/instance_verification/engine.rb:41]: s/image/instance/

File: engines/instance_verification/spec/requests/api/connect/v3/systems/products_controller_spec.rb [hunk 2]
```diff
@@ -95,6 +95,83 @@
...
+      context 'when the extension is not free' do
+        let(:product) do
+          FactoryGirl.create(
+            :product, :with_mirrored_repositories, :extension, free: false, base_products: [base_product]
+          )
+        end
+
+        it 'reports an error' do
+          data = JSON.parse(response.body)
+          expect(data['error']).to eq('Instance verification failed: The product is not available for this image')
```
Comment (rjschwei) [inline engines/instance_verification/spec/requests/api/connect/v3/systems/products_controller_spec.rb:129]: s/image/instance/

**Comment by rjschwei** [comment]
There is another problem, not with the implementation per-se, but for 3rd party CSP it is conceivable that they do sell SELS+HA. Thus by adding this into the general public cloud verification code we are forcing the model that applies to AWS, Google, and Azure onto everyone else. That is not correct. Therefore we need to be able to either enable this enforcement in the config or side load it when needed. I think a configuration option is preferable.

**Comment by ikapelyukhin** [comment]
@rjschwei sure, I'll add an opt-out option to disable this.

**Comment by suse-tests-fail** [comment]
Prophet reports failure.
https://ci.suse.de/job/scc-RMT-integration-tests/108528/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment.
Failed fast in project rmt on step docker-compose-run-tests.
Failed fast in project rmt on step docker-compose-run-tests.
Failed fast in project rmt on step docker-compose-run-tests.
Failed fast in project rmt on step docker-compose-run-tests. (Merged 028c5ac84bdfadaebdafb65a8a6965b1da53dd6b into b7219a4c03ecaf2d9dd2d5ddc3bbe466f38edc57)

**Comment by suse-tests-pass** [comment]
Well Done! Your tests are still passing.
https://ci.suse.de/job/scc-RMT-integration-tests/111310/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment. (Merged 028c5ac84bdfadaebdafb65a8a6965b1da53dd6b into 9d9cdc8f92a8c713b71060380c534c24ae7f1fa4)

**Comment by suse-tests-pass** [comment]
Well Done! Your tests are still passing.
https://ci.suse.de/job/scc-RMT-integration-tests/112070/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment. (Merged 028c5ac84bdfadaebdafb65a8a6965b1da53dd6b into 02264c8af8d25385542573785f60fe37867d4618)

**Comment by suse-tests-pass** [comment]
Well Done! Your tests are still passing.
https://ci.suse.de/job/scc-RMT-integration-tests/112590/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment. (Merged 028c5ac84bdfadaebdafb65a8a6965b1da53dd6b into b4828739a2d90bc60f1bf91ada1101157557a203)

**Comment by suse-tests-pass** [comment]
Well Done! Your tests are still passing.
https://ci.suse.de/job/scc-RMT-integration-tests/112600/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment. (Merged 028c5ac84bdfadaebdafb65a8a6965b1da53dd6b into 6effbb23e766e37ac1f25c3342b4f6ea857c85e2)

**Comment by suse-tests-pass** [comment]
Well Done! Your tests are still passing.
https://ci.suse.de/job/scc-RMT-integration-tests/116005/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment. (Merged 028c5ac84bdfadaebdafb65a8a6965b1da53dd6b into 82eff393eb69e82402b6eb3b85229d1ebf83b722)

**Comment by suse-tests-pass** [comment]
Well Done! Your tests are still passing.
https://ci.suse.de/job/scc-RMT-integration-tests/116007/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment. (Merged 028c5ac84bdfadaebdafb65a8a6965b1da53dd6b into dcfbc8ef4081a266ac343cba1d283fb6866b4400)

**Comment by suse-tests-pass** [comment]
Well Done! Your tests are still passing.
https://ci.suse.de/job/scc-RMT-integration-tests/116008/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment. (Merged 028c5ac84bdfadaebdafb65a8a6965b1da53dd6b into 5816526b08f8aa7fb867e3056ceab435473bfeee)

**Comment by suse-tests-pass** [comment]
Well Done! Your tests are still passing.
https://ci.suse.de/job/scc-RMT-integration-tests/116009/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment. (Merged 028c5ac84bdfadaebdafb65a8a6965b1da53dd6b into f1c2c37f146e1aacf819c459840d039dc05d589b)

**Comment by ikapelyukhin** [comment]
> I think a configuration option is preferable.

@rjschwei do we still want the option with the current subscription-based approach?

**Comment by suse-tests-pass** [comment]
Well Done! Your tests are still passing.
https://ci.suse.de/job/scc-RMT-integration-tests/116116/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment. (Merged 33863a65bf84f1567e2ef38ef471f00b26b44d8d into f1c2c37f146e1aacf819c459840d039dc05d589b)

**Comment by suse-tests-pass** [comment]
Well Done! Your tests are still passing.
https://ci.suse.de/job/scc-RMT-integration-tests/116146/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment. (Merged 33863a65bf84f1567e2ef38ef471f00b26b44d8d into a76145dba0e5c82a8517b740cb7a62a1b0918db4)

**Comment by suse-tests-pass** [comment]
Well Done! Your tests are still passing.
https://ci.suse.de/job/scc-RMT-integration-tests/116147/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment. (Merged 33863a65bf84f1567e2ef38ef471f00b26b44d8d into 9a0d09621993f7855a89e7792a234af56571b879)

File: engines/instance_verification/lib/instance_verification/engine.rb (hunk 1 above)
Comment (ikapelyukhin) [reply engines/instance_verification/lib/instance_verification/engine.rb]: Done.

**Comment by suse-tests-pass** [comment]
Well Done! Your tests are still passing.
https://ci.suse.de/job/scc-RMT-integration-tests/116765/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment. (Merged c43ba73a8b656f2481cb2e1ba2f19c8987a99a3d into 32cdd85a788d5cfb2cf5724888c5ba03aa231c8f)

**Comment by rjschwei** [comment]
> > I think a configuration option is preferable.
> 
> @rjschwei do we still want the option with the current subscription-based approach?

I think with the subscription based approach we are OK without having a configuration option.

**Pull request merged**